diff --git a/openssl/include/openssl/ssl.h.in b/openssl/include/openssl/ssl.h.in
index 383c5bc411..2019884e73 100644
--- a/openssl/include/openssl/ssl.h.in
+++ b/openssl/include/openssl/ssl.h.in
@@ -1922,6 +1922,8 @@ size_t SSL_client_hello_get0_compression_methods(SSL *s,
 int SSL_client_hello_get1_extensions_present(SSL *s, int **out, size_t *outlen);
 int SSL_client_hello_get_extension_order(SSL *s, uint16_t *exts,
                                          size_t *num_exts);
+int SSL_client_hello_get0_received_ext(SSL *s, uint16_t *exts,
+                                       size_t *num_exts);
 int SSL_client_hello_get0_ext(SSL *s, unsigned int type,
                               const unsigned char **out, size_t *outlen);
 
diff --git a/openssl/ssl/ssl_lib.c b/openssl/ssl/ssl_lib.c
index 9696a4c55f..beb71fabdd 100644
--- a/openssl/ssl/ssl_lib.c
+++ b/openssl/ssl/ssl_lib.c
@@ -6930,6 +6930,41 @@ int SSL_client_hello_get_extension_order(SSL *s, uint16_t *exts, size_t *num_ext
     return 1;
 }
 
+int SSL_client_hello_get0_received_ext(SSL *s, uint16_t *exts, size_t *num_exts)
+{
+    PACKET extensions;
+    size_t num = 0;
+    SSL_CONNECTION *sc = SSL_CONNECTION_FROM_SSL(s);
+
+    if (sc == NULL)
+        return 0;
+
+    if (sc->clienthello == NULL || num_exts == NULL)
+        return 0;
+
+    extensions = sc->clienthello->extensions;
+    while (PACKET_remaining(&extensions) > 0) {
+        unsigned int type;
+        PACKET extension;
+
+        if (!PACKET_get_net_2(&extensions, &type) ||
+            !PACKET_get_length_prefixed_2(&extensions, &extension)) {
+            SSLfatal(sc, SSL_AD_DECODE_ERROR, SSL_R_BAD_EXTENSION);
+            goto err;
+        }
+
+        if (exts != NULL)
+            exts[num] = (uint16_t) type;
+        num++;
+    }
+
+    *num_exts = num;
+    return 1;
+
+ err:
+    return 0;
+}
+
 int SSL_client_hello_get0_ext(SSL *s, unsigned int type, const unsigned char **out,
                        size_t *outlen)
 {
